File: DSL/de.unibonn.simpleml/src/test/kotlin/de/unibonn/simpleml/testing/TestRangesTest.kt
package de.unibonn.simpleml.testing

import de.unibonn.simpleml.location.XtextPosition
import de.unibonn.simpleml.location.XtextRange
import de.unibonn.simpleml.testing.FindTestRangesResult.CloseWithoutOpenFailure
import de.unibonn.simpleml.testing.FindTestRangesResult.OpenWithoutCloseFailure
import de.unibonn.simpleml.testing.FindTestRangesResult.Success
import de.unibonn.simpleml.testing.TestMarker.CLOSE
import de.unibonn.simpleml.testing.TestMarker.OPEN
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeInstanceOf
import org.junit.jupiter.api.Test

class TestRangesTest {

    @Test
    fun `should find all ranges enclosed by test markers in order of opening markers`() {
        val result = findTestRanges("text${OPEN}text$CLOSE\n${OPEN}text$CLOSE")
        result.shouldBeInstanceOf<Success>()
        result.ranges.shouldHaveSize(2)

        result.ranges[0] shouldBe XtextRange.fromInts(
            startLine = 1,
            startColumn = 6,
            endLine = 1,
            endColumn = 10,
            length = 4
        )

        result.ranges[1] shouldBe XtextRange.fromInts(
            startLine = 2,
            startColumn = 2,
            endLine = 2,
            endColumn = 6,
            length = 4
        )
    }

    @Test
    fun `should handle nested test markers`() {
        val result = findTestRanges("$OPEN\n    $OPEN$CLOSE\n$CLOSE")
        result.shouldBeInstanceOf<Success>()
        result.ranges.shouldHaveSize(2)

        result.ranges[0] shouldBe XtextRange.fromInts(
            startLine = 1,
            startColumn = 2,
            endLine = 3,
            endColumn = 1,
            length = 8
        )

        result.ranges[1] shouldBe XtextRange.fromInts(
            startLine = 2,
            startColumn = 6,
            endLine = 2,
            endColumn = 6,
            length = 0
        )
    }

    @Test
    fun `should handle line feed (Unix)`() {
        val result = findTestRanges("\n$OPEN\n$CLOSE")
        result.shouldBeInstanceOf<Success>()
        result.ranges.shouldHaveSize(1)

        result.ranges[0] shouldBe XtextRange.fromInts(
            startLine = 2,
            startColumn = 2,
            endLine = 3,
            endColumn = 1,
            length = 1
        )
    }

    @Test
    fun `should handle carriage return (MacOS)`() {
        val result = findTestRanges("\r$OPEN\r$CLOSE")
        result.shouldBeInstanceOf<Success>()
        result.ranges.shouldHaveSize(1)

        result.ranges[0] shouldBe XtextRange.fromInts(
            startLine = 2,
            startColumn = 2,
            endLine = 3,
            endColumn = 1,
            length = 1
        )
    }

    @Test
    fun `should handle carriage return + line feed (Windows)`() {
        val result = findTestRanges("\r\n$OPEN\r\n$CLOSE")
        result.shouldBeInstanceOf<Success>()
        result.ranges.shouldHaveSize(1)

        result.ranges[0] shouldBe XtextRange.fromInts(
            startLine = 2,
            startColumn = 2,
            endLine = 3,
            endColumn = 1,
            length = 2
        )
    }

    @Test
    fun `should report closing test markers without matching opening test marker`() {
        val result = findTestRanges("$OPEN\n$CLOSE$CLOSE")
        result.shouldBeInstanceOf<CloseWithoutOpenFailure>()
        result.position shouldBe XtextPosition.fromInts(line = 2, column = 2)
        result.message shouldBe "Found '$CLOSE' without previous '$OPEN' at 2:2."
    }

    @Test
    fun `should report opening test markers without matching closing test marker`() {
        val result = findTestRanges("$OPEN\n$OPEN$OPEN$CLOSE")
        result.shouldBeInstanceOf<OpenWithoutCloseFailure>()

        result.positions.shouldHaveSize(2)
        result.positions[0] shouldBe XtextPosition.fromInts(line = 1, column = 1)
        result.positions[1] shouldBe XtextPosition.fromInts(line = 2, column = 1)

        result.message shouldBe "Found '$OPEN' without following '$CLOSE' at 1:1, 2:1."
    }
}

File: DSL-langium/tests/helpers/testRanges.test.ts
import { describe, expect, test } from 'vitest';
import { CloseWithoutOpenError, findTestRanges, OpenWithoutCloseError } from './testRanges';
import { Position, Range } from 'vscode-languageserver';
import { CLOSE, OPEN } from './testMarker';

describe('findTestRanges', () => {
    test('should find all ranges enclosed by test markers in order of opening markers', () => {
        const result = findTestRanges(`text${OPEN}text${CLOSE}\n${OPEN}text${CLOSE}`);
        expect(result.isOk).toEqual(true);

        if (result.isOk) {
            const ranges = result.value;
            expect(ranges).toStrictEqual([Range.create(1, 6, 1, 10), Range.create(2, 2, 2, 6)]);
        }
    });

    test('should handle nested test markers', () => {
        const result = findTestRanges(`${OPEN}\n    ${OPEN}${CLOSE}\n${CLOSE}`);
        expect(result.isOk).toBeTruthy();

        if (result.isOk) {
            const ranges = result.value;
            expect(ranges).toStrictEqual([Range.create(1, 2, 3, 1), Range.create(2, 6, 2, 6)]);
        }
    });

    test('should handle line feed (Unix)', () => {
        const result = findTestRanges(`\n${OPEN}\n${CLOSE}`);
        expect(result.isOk).toBeTruthy();

        if (result.isOk) {
            const ranges = result.value;
            expect(ranges).toStrictEqual([Range.create(2, 2, 3, 1)]);
        }
    });

    test('should handle carriage return (MacOS)', () => {
        const result = findTestRanges(`\r${OPEN}\r${CLOSE}`);
        expect(result.isOk).toBeTruthy();

        if (result.isOk) {
            const ranges = result.value;
            expect(ranges).toStrictEqual([Range.create(2, 2, 3, 1)]);
        }
    });

    test('should handle carriage return + line feed (Windows)', () => {
        const result = findTestRanges(`\r\n${OPEN}\r\n${CLOSE}`);
        expect(result.isOk).toBeTruthy();

        if (result.isOk) {
            const ranges = result.value;
            expect(ranges).toStrictEqual([Range.create(2, 2, 3, 1)]);
        }
    });

    test('should report closing test markers without matching opening test marker', () => {
        const result = findTestRanges(`${OPEN}\n${CLOSE}${CLOSE}`);
        expect(result.isErr).toBeTruthy();

        if (result.isErr) {
            const error = result.error;
            expect(error).toBeInstanceOf(CloseWithoutOpenError);
            expect((error as CloseWithoutOpenError).position).toEqual(Position.create(2, 2));
            expect(error.message).toBe(`Found '${CLOSE}' without previous '${OPEN}' at 2:2.`);
        }
    });

    test('should report opening test markers without matching closing test marker', () => {
        const result = findTestRanges(`${OPEN}\n${OPEN}${OPEN}${CLOSE}`);
        expect(result.isErr).toBeTruthy();

        if (result.isErr) {
            const error = result.error;
            expect(error).toBeInstanceOf(OpenWithoutCloseError);
            expect((error as OpenWithoutCloseError).positions).toEqual([Position.create(1, 1), Position.create(2, 1)]);
            expect(error.message).toBe(`Found '${OPEN}' without following '${CLOSE}' at 1:1, 2:1.`);
        }
    });
});
